lib/active_model_serializers/adapter/json_api/resource_identifier.rb

```diff
@@ -20,11 +20,13 @@ def as_json
 
         def type_for(serializer)
           return serializer._type if serializer._type
-          if ActiveModelSerializers.config.jsonapi_resource_type == :singular
-            serializer.object.class.model_name.singular
-          else
-            serializer.object.class.model_name.plural
+          type = serializer.object.class.to_s.split('::')
+          type.map! { |t| t.underscore.downcase }
+          type = type.join ActiveModelSerializers.config.jsonapi_namespace_separator
```

Comment:
I was thinking this would be a good use-case for gsub, but it looks like that is would be marginally slower and a little harder to read

```ruby
Benchmark.realtime { 1000.times { 'AdminUser::Person::THING'.gsub(/[a-zA-Z]+(::|\z)/) do |sub| sub.underscore.downcase.sub('/','--') end } }
# => 0.07714433899673168
>> Benchmark.realtime { 1000.times { 'AdminUser::Person::THING'.split('::').map! {|t| t.underscore.downcase }.join('--') } }     
# => 0.07239663798827678
```